core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
+      getTransactionState(transactionalId) match {
+        case Left(err) =>
+          responseCallback(err)
+
+        case Right(Some(epochAndMetadata)) =>
+          val metadata = epochAndMetadata.transactionMetadata
+
+          metadata synchronized {
+            if (epochAndMetadata.coordinatorEpoch != coordinatorEpoch) {
+              // the coordinator epoch has changed, reply to client immediately with with NOT_COORDINATOR
+              responseCallback(Errors.NOT_COORDINATOR)
+            } else {
+              // do not need to check the metadata object itself since no current thread should be able to modify it
+              // under the same coordinator epoch, so directly append to txn log now
+
+              replicaManager.appendRecords(
```

Comment:
Hmm, it's a bit tricky. We need to hold the lock until local append returns, but not when calling `delayedProducePurgatory.tryCompleteElseWatch`.

BTW I have been thinking about this race condition a bit more, and now I feel it is actually okay to not lock the memory cache while appending to local log, and here is my reasoning: When there is an emigration / immigration happening after the txn metadata object is checked but before the local log is appended, then the following case can happen:

1. `handleFollower` is received and handled, in which the local log append will be truncated; the delayed produce will eventually be timed out and an error returned to the user.
2. `handleFollower` is not received, and `handleLeader` with new epoch is received and handled; the append log will be appended and epoch check will fail in the callback, and a `NOT_COORDINATOR` will be sent back to client and it will retry; we will eventually get an idempotent append in the txn log which is okay for loading.
3. none of `handleFollower` or `handleLeader` is received, and the append is successful, it is locally the same as the event happens after the append to the txn log.

WDTY?